### PR TITLE
Adding e2e test using Playwright

### PR DESCRIPTION
**Pull request by salman90**
# Pull Request

## Purpose

*  Adding e2e test using Playwright to test the wrapper against AAD and B2C.

## Does this introduce a breaking change

```console
    [ ] Yes
    [ X ] No
```

## Pull request type

What kind of change does this Pull Request introduce?

```console
    [ ] Bugfix
    [ X ] Feature
    [ ] Code style update (formatting, local variables)
    [ ] Documentation content changes
    [ ] Other... Please describe:
```

## How to test

* Get the code

```console
    git clone [repo-address]
    cd [repo-name]
    git checkout [branch-name]
    npm install
```

## What to check

ex: verify that the following are valid:

* ...

### CODE REVIEW

File: demo/test/e2eTestUtils/NodeCacheTestUtils.ts
Comment (derisen) [inline demo/test/e2eTestUtils/NodeCacheTestUtils.ts:1]: let's put this into e2eTestUtils.ts

File: demo/App/app.js [hunk 1]
```diff
@@ -15,45 +17,61 @@ const router = require('./routes/router');
...
- */
-app.use(session({
-    secret: 'ENTER_YOUR_SECRET_HERE',
-    resave: false,
-    saveUninitialized: false,
-    cookie: {
-        secure: false,
-    }
-}));
+async function main(msid = null) {
```
Comment (derisen) [inline demo/App/app.js:20]: do you need this default value? can't you just check undefined below?

File: demo/test/e2eTestUtils/NodeCacheTestUtils.ts
Comment (derisen) [reply demo/test/e2eTestUtils/NodeCacheTestUtils.ts:1]: i take it back

File: demo/test/e2eTestUtils/constants.ts [hunk 2]
```diff
@@ -0,0 +1,5 @@
+export const CREDENTIAL_TYPES = {
```
Comment (derisen) [inline demo/test/e2eTestUtils/constants.ts:1]: Can we put this into NodeCacheTestUtils.ts instead?

File: demo/test/aad.spec.ts [hunk 3]
```diff
@@ -4,115 +4,293 @@
...
-    test.afterEach(async () => {
-      await page.close();
-      await context.close();
-    });
+        test.beforeAll(async () => {
+            server = app.listen(port);
+            //msid = new MsIdExpress.WebAppAuthClientBuilder(appSettings).build();
+            msid = await new MsIdExpress.WebAppAuthClientBuilder(appSettings)
+                .withKeyVaultCredentials({
+                    credentialType: process.env['CREDENTIAL_TYPE'],
```
Comment (derisen) [inline demo/test/aad.spec.ts:74]: Are these in the repo secrets?

File: demo/test/aad.spec.ts [hunk 4]
```diff
@@ -4,115 +4,293 @@
...
+            await screenshot.takeScreenshot(page, 'samplePagePostLogin');
+            page.click('#groupsButton');
+            await page.waitForFunction(`window.location.href.startsWith("${SAMPLE_HOME_URL}")`);
+            await expect(page.locator(`text=unauthorized`).first()).toBeVisible();
+
+            const cachedTokens = NodeCacheTestUtils.getTokens(tokenCache);
+            expect(cachedTokens.accessTokens.length).toBe(1);
+            expect(cachedTokens.idTokens.length).toBe(1);
+            expect(cachedTokens.refreshTokens.length).toBe(1);
+        });
```
Comment (derisen) [inline demo/test/aad.spec.ts:270]: let's put these rbac related tests into a separate file

File: package.json [hunk 5]
```diff
@@ -45,6 +45,7 @@
     "eslint-plugin-security": "^1.7.1",
     "husky": "^6.0.0",
     "jest": "^27.5.1",
+    "playwright-core": "^1.31.2",
```
Comment (derisen) [inline package.json:48]: do we actually need this dependency in the main package.json?

Comment (salman90) [reply package.json]: Removed

File: demo/test/aad.spec.ts [hunk 6]
```diff
@@ -4,115 +4,293 @@
...
-      await page.waitForFunction(`window.location.href.startsWith("${SAMPLE_HOME_URL}")`);
-      await screenshot.takeScreenshot(page, "samplePagePostLogout");
-      await expect(page.locator(`text=${username}`).first()).not.toBeVisible();
-      await expect(page.locator(`text=Sign-in to access your resources`).first()).toBeVisible();
+            await clickSignOut(page, screenshot);
+            await page.waitForFunction(`window.location.href.startsWith("${SAMPLE_HOME_URL}")`);
+            await screenshot.takeScreenshot(page, 'samplePagePostLogout');
+            await expect(page.locator(`text=${username}`).first()).not.toBeVisible();
+            await expect(page.locator(`text=Sign-in to access your resources`).first()).toBeVisible();
+        });
```
[the remark this answers is not in the document]
Comment (salman90) [reply demo/test/aad.spec.ts]: Done

File: demo/test/aad.spec.ts (hunk 3 above)
Comment (salman90) [reply demo/test/aad.spec.ts]: Added

File: demo/test/e2eTestUtils/constants.ts (hunk 2 above)
Comment (salman90) [reply demo/test/e2eTestUtils/constants.ts]: done

File: demo/App/app.js (hunk 1 above)
Comment (salman90) [reply demo/App/app.js]: Done

**Review by derisen — APPROVED** [review]
LGTM!

**Pull request merged**